File: 2049_count_highest_score_nodes/count_highest_score_nodes.go
/*
 * @Date: 2022-03-10 23:48:16
 * @Author: Mengsen Wang
 * @LastEditors: Mengsen Wang
 * @LastEditTime: 2022-03-11 00:42:28
 * @FilePath: /algorithm/2049_count_highest_score_nodes/count_highest_score_nodes.go
 */

package main

func countHighestScoreNodes(parents []int) (ans int) {
	n := len(parents)
	children := make([][]int, n)
	for node := 1; node < n; node++ {
		p := parents[node]
		children[p] = append(children[p], node)
	}

	maxScore := 0
	var dfs func(int) int
	dfs = func(node int) int {
		score, size := 1, n-1
		for _, ch := range children[node] {
			sz := dfs(ch)
			score *= sz
			size -= sz
		}
		if node > 0 {
			score *= size
		}
		if score == maxScore {
			ans++
		} else if score > maxScore {
			maxScore = score
			ans = 1
		}
		return n - size
	}
	dfs(0)
	return
}

func main() {
	assert := func(b bool) {
		if !b {
			panic("Not Passed")
		}
	}

	assert(countHighestScoreNodes([]int{-1, 2, 0, 2, 0}) == 3)
	assert(countHighestScoreNodes([]int{-1, 2, 0}) == 2)
}

File: cpp/2049_count_highest_score_nodes/count_highest_score_nodes.cpp
/*
 * @Date: 2022-03-10 23:48:09
 * @Author: Mengsen Wang
 * @LastEditors: Mengsen Wang
 * @LastEditTime: 2022-03-11 00:39:18
 * @FilePath: /algorithm/2049_count_highest_score_nodes/count_highest_score_nodes.cpp
 * /algorithm/2049_count_highest_score_nodes/count_highest_score_nodes.cpp
 */

#include <cassert>
#include <vector>

using namespace std;

class Solution {
 public:
  long maxScore = 0;
  int cnt = 0;
  int n;
  vector<vector<int>> children;

  int dfs(int node) {
    long score = 1;
    int size = n - 1;
    for (int c : children[node]) {
      int t = dfs(c);
      score *= t;
      size -= t;
    }
    if (node != 0) {
      score *= size;
    }
    if (score == maxScore) {
      cnt++;
    } else if (score > maxScore) {
      maxScore = score;
      cnt = 1;
    }
    return n - size;
  }

  int countHighestScoreNodes(vector<int>&& parents) {
    this->n = parents.size();
    this->children = vector<vector<int>>(n);
    for (int i = 0; i < n; i++) {
      int p = parents[i];
      if (p != -1) {
        children[p].emplace_back(i);
      }
    }
    dfs(0);
    return cnt;
  }
};

int main() {
  assert(Solution().countHighestScoreNodes({-1, 2, 0, 2, 0}) == 3);
  assert(Solution().countHighestScoreNodes({-1, 2, 0}) == 2);
}